pypmj/optimizer.py

```diff
@@ -0,0 +1,143 @@
+"""Defines the Optimizer class which acts as an interface between the jcmwave.optimizer module
```

Comment:
Class to interface JCMsuite's optimizer module. Minimal example to optimize the Q factor of certain cavity:
```py
import numpy as np
import pypmj as jpy

def objective_func(simulation):
    if not 'eigenmode' in simulation.jcm_results[0]['eigenvalues']:
        return None

    if not len(simulation.jcm_results[0]['eigenvalues']['eigenmode']):
        return None

    eigenvalue = max(simulation.jcm_results[0]['eigenvalues']['eigenmode'], key = lambda x: np.imag(x) if np.imag(x) < 0 else -1e20)
    q_factor = .5 * np.abs(np.real(eigenvalue) / np.imag(eigenvalue))

    return -q_factor

jpy.load_config_file('pypmj_config.cfg')
jpy.import_jcmwave()

constant_keys = {
    #...
}
domain = [
    #...
]
constraints = [
    #...
]

project = jpy.JCMProject('TestProject', project_file_name = 'project.jcmpt')
optimizer = jpy.Optimizer(project, domain, constraints, constant_keys, max_iter=100)
optimizer.run(objective_func)

optimizer.logger.info('Maximal q factor Q = {} found for parameters '.format(-optimizer.study.info()['min_objective']))
print(optimizer.study.info()['min_params'])
```